iree/compiler/Dialect/Flow/Transforms/DeduplicateExecutables.cpp

```diff
@@ -0,0 +1,208 @@
+        if (!areRegionsEquivalent(&lhsRegions[i], &rhsRegions[i])) {
+          return false;
+        }
+      }
+    }
+  }
+
+  return true;
+}
+
+bool areExecutablesEquivalent(ExecutableOp lhs, ExecutableOp rhs) {
+  auto lhsModule = lhs.getInnerModule();
+  auto rhsModule = rhs.getInnerModule();
+
+  // Must have the same number of entry point ops, with the same attributes.
+  // Entry point op symbol names are expected to differ, that won't affect
+  // equivalence.
+  auto lhsEntryOps = llvm::to_vector<1>(lhsModule.getOps<DispatchEntryOp>());
+  auto rhsEntryOps = llvm::to_vector<1>(rhsModule.getOps<DispatchEntryOp>());
+  if (lhsEntryOps.size() != rhsEntryOps.size()) {
+    return false;
+  }
+  for (int i = 0; i < lhsEntryOps.size(); ++i) {
+    if (lhsEntryOps[i].getAttrs() != rhsEntryOps[i].getAttrs()) {
+      return false;
+    }
+  }
+
+  // Must have the same number of functions, with each listed in the same order
+  // and with equivalent regions inside.
+  auto lhsFuncOps = llvm::to_vector<1>(lhsModule.getOps<FuncOp>());
+  auto rhsFuncOps = llvm::to_vector<1>(rhsModule.getOps<FuncOp>());
+  if (lhsFuncOps.size() != rhsFuncOps.size()) {
+    return false;
+  }
+  for (int i = 0; i < lhsFuncOps.size(); ++i) {
+    auto lhsRegion = lhsFuncOps[i].getCallableRegion();
+    auto rhsRegion = rhsFuncOps[i].getCallableRegion();
+    return areRegionsEquivalent(lhsRegion, rhsRegion);
+  }
+
+  return true;
+}
+
+}  // namespace
+
+class DeduplicateExecutablesPass
+    : public PassWrapper<DeduplicateExecutablesPass, OperationPass<ModuleOp>> {
+ public:
+  explicit DeduplicateExecutablesPass() {}
+  DeduplicateExecutablesPass(const DeduplicateExecutablesPass &pass) {}
+
+  void runOnOperation() override {
+    auto moduleOp = getOperation();
+    auto executableOps = llvm::to_vector<8>(moduleOp.getOps<ExecutableOp>());
+    auto builder = OpBuilder::atBlockBegin(moduleOp.getBody());
+
+    SmallVector<ExecutableOp, 3> duplicateExecutableOps;
+    DenseMap<Attribute, Attribute> entryPointRefReplacements;
+
+    // For each executable, find the first executable which it is equivalent to.
+    for (int i = executableOps.size() - 1; i >= 0; --i) {
+      auto executableOp = executableOps[i];
+
+      for (int j = 0; j < i; ++j) {
+        auto comparisonExecutableOp = executableOps[j];
+
+        if (!areExecutablesEquivalent(executableOp, comparisonExecutableOp)) {
+          continue;
+        }
+
+        // Found an equivalent executable! Record it and move on to the next.
+        duplicateExecutableOps.push_back(executableOp);
+
+        // Record entry point reference replacements.
+        auto dispatchEntryOps = llvm::to_vector<1>(
+            executableOp.getBlock().getOps<DispatchEntryOp>());
+        auto comparisonDispatchEntryOps = llvm::to_vector<1>(
+            comparisonExecutableOp.getBlock().getOps<DispatchEntryOp>());
+        assert(dispatchEntryOps.size() == comparisonDispatchEntryOps.size());
```

Comment:
llvm::zip this - then you don't need to get these into vectors, use a k loop variable, etc